File: Reverse Parenthesis/Reverse Parenthesis.py
# Decodes a string by reversing substrings inside parentheses. For each closing parenthesis, the substring inside the matching parentheses is reversed in place.
# Space Complexity: O(n)
# Time Complexity: O(n^2)
def decode(s):
    stack = []

    # Iterate through each character in the input string
    for char in s:
        # If the character is a closing parenthesis
        if char == ')':
            # Prepare to collect characters to be reversed
            temp = []
            # Pop characters from the stack until an opening parenthesis is found
            while stack and stack[-1] != '(': 
                temp.append(stack.pop())  # Add popped character to temp
            stack.pop()  # Remove the opening parenthesis '(' from the stack
            # Push the reversed substring (inside parentheses) back onto the stack
            stack.extend(temp)
        else:
            # If not a closing parenthesis, push the character onto the stack
            stack.append(char)

    return ''.join(stack)

File: Reverse Parenthesis/Reverse Parenthesis.rs
// Time Complexity: O(n^2)
// Space Complexity: O(n)
// Decodes a string by reversing substrings inside parentheses. For each closing parenthesis,
// the substring inside the matching parentheses is reversed in place.
fn decode(s: &str) -> String {
    let mut stack: Vec<char> = Vec::new();

    // Iterate through each character in the input string
    for c in s.chars() {
        // If the character is a closing parenthesis
        if c == ')' {
            let mut temp = Vec::new();
            // Pop characters from the stack until an opening parenthesis is found
            while let Some(&top) = stack.last() {
                // If the top of the stack is an opening parenthesis, remove it and break
                if top == '(' {
                    stack.pop(); // remove '('
                    break;
                }
                // Otherwise, pop the character and add it to temp
                temp.push(stack.pop().unwrap());
            }
            // Push the reversed substring (inside parentheses) back onto the stack
            stack.extend(temp);
        } else {
            // If not a closing parenthesis, push the character onto the stack
            stack.push(c);
        }
    }

    // Collect the characters from the stack into a string and return
    stack.iter().collect()
}

